#import of libraries


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
#Reading of Dataset
data=pd.read_csv('Leads.csv')

In [4]:
data.shape

(9240, 37)

In [5]:
# Drop Duplicate Values
data =data.drop_duplicates()

In [6]:
data.shape

(9240, 37)

In [7]:
# Glimpse of data set
data.head()

Prospect ID  Lead Number              Lead Origin  \
0  7927b2df-8bba-4d29-b9a2-b6e0beafe620       660737                      API   
1  2a272436-5132-4136-86fa-dcc88c88f482       660728                      API   
2  8cc8c611-a219-4f35-ad23-fdfd2656bd8a       660727  Landing Page Submission   
3  0cc2df48-7cf4-4e39-9de9-19797f9b38cc       660719  Landing Page Submission   
4  3256f628-e534-4826-9d63-4a8b88782852       660681  Landing Page Submission   

      Lead Source Do Not Email Do Not Call  Converted  TotalVisits  \
0      Olark Chat           No          No          0          0.0   
1  Organic Search           No          No          0          5.0   
2  Direct Traffic           No          No          1          2.0   
3  Direct Traffic           No          No          0          1.0   
4          Google           No          No          1          2.0   

   Total Time Spent on Website  Page Views Per Visit  ...  \
0                            0                   0.0  ...   
1                          674                   2.5  ...   
2                         1532                   2.0  ...   
3                          305                   1.0  ...   
4                         1428                   1.0  ...   

  Get updates on DM Content    Lead Profile    City  \
0                        No          Select  Select   
1                        No          Select  Select   
2                        No  Potential Lead  Mumbai   
3                        No          Select  Mumbai   
4                        No          Select  Mumbai   

  Asymmetrique Activity Index Asymmetrique Profile Index  \
0                   02.Medium                  02.Medium   
1                   02.Medium                  02.Medium   
2                   02.Medium                    01.High   
3                   02.Medium                    01.High   
4                   02.Medium                    01.High   

  Asymmetrique Activity Score Asymmetrique Profile Score  \
0                        15.0                       15.0   
1                        15.0                       15.0   
2                        14.0                       20.0   
3                        13.0                       17.0   
4                        15.0                       18.0   

  I agree to pay the amount through cheque  \
0                                       No   
1                                       No   
2                                       No   
3                                       No   
4                                       No   

  A free copy of Mastering The Interview Last Notable Activity  
0                                     No              Modified  
1                                     No          Email Opened  
2                                    Yes          Email Opened  
3                                     No              Modified  
4                                     No              Modified  

[5 rows x 37 columns]

In [8]:
# Shape and size of dataset
print(data.shape,data.size)

(9240, 37) 341880


# Data Cleaning

In [9]:
# checking for the null values
data.isnull().sum()

Prospect ID                                         0
Lead Number                                         0
Lead Origin                                         0
Lead Source                                        36
Do Not Email                                        0
Do Not Call                                         0
Converted                                           0
TotalVisits                                       137
Total Time Spent on Website                         0
Page Views Per Visit                              137
Last Activity                                     103
Country                                          2461
Specialization                                   1438
How did you hear about X Education               2207
What is your current occupation                  2690
What matters most to you in choosing a course    2709
Search                                              0
Magazine                                            0
Newspaper Article                                   0
X Education Forums                                  0
Newspaper                                           0
Digital Advertisement                               0
Through Recommendations                             0
Receive More Updates About Our Courses              0
Tags                                             3353
Lead Quality                                     4767
Update me on Supply Chain Content                   0
Get updates on DM Content                           0
Lead Profile                                     2709
City                                             1420
Asymmetrique Activity Index                      4218
Asymmetrique Profile Index                       4218
Asymmetrique Activity Score                      4218
Asymmetrique Profile Score                       4218
I agree to pay the amount through cheque            0
A free copy of Mastering The Interview              0
Last Notable Activity                               0
dtype: int64

In [10]:
# Finding the columns with null values
data.columns[data.isnull().any()]

Index(['Lead Source', 'TotalVisits', 'Page Views Per Visit', 'Last Activity',
       'Country', 'Specialization', 'How did you hear about X Education',
       'What is your current occupation',
       'What matters most to you in choosing a course', 'Tags', 'Lead Quality',
       'Lead Profile', 'City', 'Asymmetrique Activity Index',
       'Asymmetrique Profile Index', 'Asymmetrique Activity Score',
       'Asymmetrique Profile Score'],
      dtype='object')

# Separation of null values and non null values

In [14]:

non_null_data=data[data.columns[~data.isnull().any()]]

In [15]:
non_null_data.isnull().sum()

Prospect ID                                 0
Lead Number                                 0
Lead Origin                                 0
Do Not Email                                0
Do Not Call                                 0
Converted                                   0
Total Time Spent on Website                 0
Search                                      0
Magazine                                    0
Newspaper Article                           0
X Education Forums                          0
Newspaper                                   0
Digital Advertisement                       0
Through Recommendations                     0
Receive More Updates About Our Courses      0
Update me on Supply Chain Content           0
Get updates on DM Content                   0
I agree to pay the amount through cheque    0
A free copy of Mastering The Interview      0
Last Notable Activity                       0
dtype: int64

In [16]:
# make dataframe of the null values
null_data=data[data.columns[data.isnull().any()]]

In [17]:
null_data.head()

Lead Source  TotalVisits  Page Views Per Visit            Last Activity  \
0      Olark Chat          0.0                   0.0  Page Visited on Website   
1  Organic Search          5.0                   2.5             Email Opened   
2  Direct Traffic          2.0                   2.0             Email Opened   
3  Direct Traffic          1.0                   1.0              Unreachable   
4          Google          2.0                   1.0        Converted to Lead   

  Country           Specialization How did you hear about X Education  \
0     NaN                   Select                             Select   
1   India                   Select                             Select   
2   India  Business Administration                             Select   
3   India    Media and Advertising                      Word Of Mouth   
4   India                   Select                              Other   

  What is your current occupation  \
0                      Unemployed   
1                      Unemployed   
2                         Student   
3                      Unemployed   
4                      Unemployed   

  What matters most to you in choosing a course  \
0                       Better Career Prospects   
1                       Better Career Prospects   
2                       Better Career Prospects   
3                       Better Career Prospects   
4                       Better Career Prospects   

                                  Tags      Lead Quality    Lead Profile  \
0          Interested in other courses  Low in Relevance          Select   
1                              Ringing               NaN          Select   
2  Will revert after reading the email          Might be  Potential Lead   
3                              Ringing          Not Sure          Select   
4  Will revert after reading the email          Might be          Select   

     City Asymmetrique Activity Index Asymmetrique Profile Index  \
0  Select                   02.Medium                  02.Medium   
1  Select                   02.Medium                  02.Medium   
2  Mumbai                   02.Medium                    01.High   
3  Mumbai                   02.Medium                    01.High   
4  Mumbai                   02.Medium                    01.High   

   Asymmetrique Activity Score  Asymmetrique Profile Score  
0                         15.0                        15.0  
1                         15.0                        15.0  
2                         14.0                        20.0  
3                         13.0                        17.0  
4                         15.0                        18.0

In [18]:
# Finding the columns with more then 28 % null values
null_data.columns[null_data.isnull().mean()>0.28]

Index(['What is your current occupation',
       'What matters most to you in choosing a course', 'Tags', 'Lead Quality',
       'Lead Profile', 'Asymmetrique Activity Index',
       'Asymmetrique Profile Index', 'Asymmetrique Activity Score',
       'Asymmetrique Profile Score'],
      dtype='object')

In [19]:
# Droping the columns with null values more the 28 percent
null_data.drop(['What is your current occupation',
       'What matters most to you in choosing a course', 'Tags', 'Lead Quality',
       'Lead Profile', 'Asymmetrique Activity Index',
       'Asymmetrique Profile Index', 'Asymmetrique Activity Score',
       'Asymmetrique Profile Score'],axis=1,inplace=True)

<ipython-input-19-39deda09f993>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  null_data.drop(['What is your current occupation',


In [20]:
null_data.isnull().sum()

Lead Source                             36
TotalVisits                            137
Page Views Per Visit                   137
Last Activity                          103
Country                               2461
Specialization                        1438
How did you hear about X Education    2207
City                                  1420
dtype: int64

In [21]:
null_data['Country'].value_counts()

Country
India                   6492
United States             69
United Arab Emirates      53
Singapore                 24
Saudi Arabia              21
United Kingdom            15
Australia                 13
Qatar                     10
Hong Kong                  7
Bahrain                    7
Oman                       6
France                     6
unknown                    5
South Africa               4
Nigeria                    4
Germany                    4
Kuwait                     4
Canada                     4
Sweden                     3
China                      2
Asia/Pacific Region        2
Uganda                     2
Bangladesh                 2
Italy                      2
Belgium                    2
Netherlands                2
Ghana                      2
Philippines                2
Russia                     1
Switzerland                1
Vietnam                    1
Denmark                    1
Tanzania                   1
Liberia                    1
Malaysia                   1
Kenya                      1
Sri Lanka                  1
Indonesia                  1
Name: count, dtype: int64

In [22]:
# Since country is Biased so we drop the country columns
null_data.drop('Country',axis=1,inplace=True)

<ipython-input-22-ef2143fd06a0>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  null_data.drop('Country',axis=1,inplace=True)


In [23]:
col =['Specialization','How did you hear about X Education','City']
for i in col:
  print(i,null_data[i].unique())
  print(null_data[i].value_counts(normalize=True) * 100)
  print('\n')

Specialization ['Select' 'Business Administration' 'Media and Advertising' nan
 'Supply Chain Management' 'IT Projects Management' 'Finance Management'
 'Travel and Tourism' 'Human Resource Management' 'Marketing Management'
 'Banking, Investment And Insurance' 'International Business' 'E-COMMERCE'
 'Operations Management' 'Retail Management' 'Services Excellence'
 'Hospitality Management' 'Rural and Agribusiness' 'Healthcare Management'
 'E-Business']
Specialization
Select                               24.891054
Finance Management                   12.509613
Human Resource Management            10.869008
Marketing Management                 10.740836
Operations Management                 6.447065
Business Administration               5.165342
IT Projects Management                4.691105
Supply Chain Management               4.473212
Banking, Investment And Insurance     4.332223
Travel and Tourism                    2.601897
Media and Advertising                 2.601897
Internation

In [24]:
# In col ['Specialization','How did you hear about X Education','City'] the percentage of select is greater than 25 % let's drop the column

In [25]:
null_data.drop(col,axis=1,inplace=True)

<ipython-input-25-49dbc5bf729c>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  null_data.drop(col,axis=1,inplace=True)


In [26]:
null_data.isnull().sum()

Lead Source              36
TotalVisits             137
Page Views Per Visit    137
Last Activity           103
dtype: int64

In [27]:
# Since their is no null values in the null_data let's concat it with the original data
data=pd.concat([non_null_data,null_data],axis=1)

In [28]:
print(non_null_data.shape,null_data.shape,data.shape)

(9240, 20) (9240, 4) (9240, 24)


In [29]:
data.columns

Index(['Prospect ID', 'Lead Number', 'Lead Origin', 'Do Not Email',
       'Do Not Call', 'Converted', 'Total Time Spent on Website', 'Search',
       'Magazine', 'Newspaper Article', 'X Education Forums', 'Newspaper',
       'Digital Advertisement', 'Through Recommendations',
       'Receive More Updates About Our Courses',
       'Update me on Supply Chain Content', 'Get updates on DM Content',
       'I agree to pay the amount through cheque',
       'A free copy of Mastering The Interview', 'Last Notable Activity',
       'Lead Source', 'TotalVisits', 'Page Views Per Visit', 'Last Activity'],
      dtype='object')

In [30]:
for i in data.columns:
  print(i,data[i].unique())
  print(data[i].value_counts(normalize=True) * 100)
  print('\n')

Prospect ID ['7927b2df-8bba-4d29-b9a2-b6e0beafe620'
 '2a272436-5132-4136-86fa-dcc88c88f482'
 '8cc8c611-a219-4f35-ad23-fdfd2656bd8a' ...
 'aac550fe-a586-452d-8d3c-f1b62c94e02c'
 '5330a7d1-2f2b-4df4-85d6-64ca2f6b95b9'
 '571b5c8e-a5b2-4d57-8574-f2ffb06fdeff']
Prospect ID
7927b2df-8bba-4d29-b9a2-b6e0beafe620    0.010823
22e9d4ef-d294-4ebf-81c7-7c7a1105aeea    0.010823
46befc49-253a-419b-abea-2fd978d2e2b1    0.010823
9d35a2c2-09d8-439f-9875-0e8bbf267f5a    0.010823
f0de9371-4dc2-48c2-9785-a08d6fc4fcb5    0.010823
                                          ...   
ff1f7582-cb7b-4b94-9cdc-3d0d0afdd9a3    0.010823
644099a2-3da4-4d23-9546-7676340a372b    0.010823
2a093175-415b-4321-9e69-ed8d9df65a3c    0.010823
c66249a3-8500-4c66-a511-312d914573de    0.010823
571b5c8e-a5b2-4d57-8574-f2ffb06fdeff    0.010823
Name: proportion, Length: 9240, dtype: float64


Lead Number [660737 660728 660727 ... 579545 579538 579533]
Lead Number
660737    0.010823
603303    0.010823
602561    0.010823
602557    0.01

In [31]:
# Since the columns ['Prospect ID','Lead Number','Do Not Call','Search','Magazine',
# 'Newspaper Article','X Education Forums','Newspaper','Digital Advertisement','Through Recommendations','Receive More Updates About Our Courses'
# 'Update me on Supply Chain Content','Get updates on DM Content','I agree to pay the amount through cheque']
# have biased data so lets drop it

data.drop(['Prospect ID','Lead Number','Do Not Call','Search','Magazine',
           'Newspaper Article','X Education Forums','Newspaper','Digital Advertisement','Through Recommendations',
           'Receive More Updates About Our Courses',
          'Update me on Supply Chain Content','Get updates on DM Content','I agree to pay the amount through cheque'],axis=1,inplace =True)

In [32]:
data.shape

(9240, 10)

In [33]:
data.columns

Index(['Lead Origin', 'Do Not Email', 'Converted',
       'Total Time Spent on Website', 'A free copy of Mastering The Interview',
       'Last Notable Activity', 'Lead Source', 'TotalVisits',
       'Page Views Per Visit', 'Last Activity'],
      dtype='object')

In [34]:
data.head()

Lead Origin Do Not Email  Converted  \
0                      API           No          0   
1                      API           No          0   
2  Landing Page Submission           No          1   
3  Landing Page Submission           No          0   
4  Landing Page Submission           No          1   

   Total Time Spent on Website A free copy of Mastering The Interview  \
0                            0                                     No   
1                          674                                     No   
2                         1532                                    Yes   
3                          305                                     No   
4                         1428                                     No   

  Last Notable Activity     Lead Source  TotalVisits  Page Views Per Visit  \
0              Modified      Olark Chat          0.0                   0.0   
1          Email Opened  Organic Search          5.0                   2.5   
2          Email Opened  Direct Traffic          2.0                   2.0   
3              Modified  Direct Traffic          1.0                   1.0   
4              Modified          Google          2.0                   1.0   

             Last Activity  
0  Page Visited on Website  
1             Email Opened  
2             Email Opened  
3              Unreachable  
4        Converted to Lead

In [35]:
# Now let's convert the bool(Yes / No)  into the numerical (1 and 0)
data['Do Not Email']=data['Do Not Email'].map({'Yes':1,'No':0})
data['A free copy of Mastering The Interview']=data['A free copy of Mastering The Interview'].map({'Yes':1,'No':0})

In [36]:
data.head()

Lead Origin  Do Not Email  Converted  \
0                      API             0          0   
1                      API             0          0   
2  Landing Page Submission             0          1   
3  Landing Page Submission             0          0   
4  Landing Page Submission             0          1   

   Total Time Spent on Website  A free copy of Mastering The Interview  \
0                            0                                       0   
1                          674                                       0   
2                         1532                                       1   
3                          305                                       0   
4                         1428                                       0   

  Last Notable Activity     Lead Source  TotalVisits  Page Views Per Visit  \
0              Modified      Olark Chat          0.0                   0.0   
1          Email Opened  Organic Search          5.0                   2.5   
2          Email Opened  Direct Traffic          2.0                   2.0   
3              Modified  Direct Traffic          1.0                   1.0   
4              Modified          Google          2.0                   1.0   

             Last Activity  
0  Page Visited on Website  
1             Email Opened  
2             Email Opened  
3              Unreachable  
4        Converted to Lead

In [37]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 10 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   Lead Origin                             9240 non-null   object 
 1   Do Not Email                            9240 non-null   int64  
 2   Converted                               9240 non-null   int64  
 3   Total Time Spent on Website             9240 non-null   int64  
 4   A free copy of Mastering The Interview  9240 non-null   int64  
 5   Last Notable Activity                   9240 non-null   object 
 6   Lead Source                             9204 non-null   object 
 7   TotalVisits                             9103 non-null   float64
 8   Page Views Per Visit                    9103 non-null   float64
 9   Last Activity                           9137 non-null   object 
dtypes: float64(2), int64(4), object(4)
memory usage: 722.0+ KB


In [38]:

# In null_data DataFrame the percentage of null values in the rows
# (Lead sores,TotalVisits,Page Views Per Visit,Last Activity) are allmost 1.5 % so let's drop the rows
column=['Lead Source','TotalVisits','Page Views Per Visit','Last Activity']

data = data.dropna(subset=column)
data.isnull().sum()


Lead Origin                               0
Do Not Email                              0
Converted                                 0
Total Time Spent on Website               0
A free copy of Mastering The Interview    0
Last Notable Activity                     0
Lead Source                               0
TotalVisits                               0
Page Views Per Visit                      0
Last Activity                             0
dtype: int64

In [39]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9074 entries, 0 to 9239
Data columns (total 10 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   Lead Origin                             9074 non-null   object 
 1   Do Not Email                            9074 non-null   int64  
 2   Converted                               9074 non-null   int64  
 3   Total Time Spent on Website             9074 non-null   int64  
 4   A free copy of Mastering The Interview  9074 non-null   int64  
 5   Last Notable Activity                   9074 non-null   object 
 6   Lead Source                             9074 non-null   object 
 7   TotalVisits                             9074 non-null   float64
 8   Page Views Per Visit                    9074 non-null   float64
 9   Last Activity                           9074 non-null   object 
dtypes: float64(2), int64(4), object(4)
memory usage: 779.8+ KB


In [40]:
data.head()

Lead Origin  Do Not Email  Converted  \
0                      API             0          0   
1                      API             0          0   
2  Landing Page Submission             0          1   
3  Landing Page Submission             0          0   
4  Landing Page Submission             0          1   

   Total Time Spent on Website  A free copy of Mastering The Interview  \
0                            0                                       0   
1                          674                                       0   
2                         1532                                       1   
3                          305                                       0   
4                         1428                                       0   

  Last Notable Activity     Lead Source  TotalVisits  Page Views Per Visit  \
0              Modified      Olark Chat          0.0                   0.0   
1          Email Opened  Organic Search          5.0                   2.5   
2          Email Opened  Direct Traffic          2.0                   2.0   
3              Modified  Direct Traffic          1.0                   1.0   
4              Modified          Google          2.0                   1.0   

             Last Activity  
0  Page Visited on Website  
1             Email Opened  
2             Email Opened  
3              Unreachable  
4        Converted to Lead

In [41]:
# Treatment of outliers
num_var=data.select_dtypes(include=['int64','float64'])
num_var.head()

Do Not Email  Converted  Total Time Spent on Website  \
0             0          0                            0   
1             0          0                          674   
2             0          1                         1532   
3             0          0                          305   
4             0          1                         1428   

   A free copy of Mastering The Interview  TotalVisits  Page Views Per Visit  
0                                       0          0.0                   0.0  
1                                       0          5.0                   2.5  
2                                       1          2.0                   2.0  
3                                       0          1.0                   1.0  
4                                       0          2.0                   1.0

In [42]:
cat_var=data.select_dtypes(include=['object'])

In [43]:
cat_var.head()

Lead Origin Last Notable Activity     Lead Source  \
0                      API              Modified      Olark Chat   
1                      API          Email Opened  Organic Search   
2  Landing Page Submission          Email Opened  Direct Traffic   
3  Landing Page Submission              Modified  Direct Traffic   
4  Landing Page Submission              Modified          Google   

             Last Activity  
0  Page Visited on Website  
1             Email Opened  
2             Email Opened  
3              Unreachable  
4        Converted to Lead

In [44]:
cat_var=pd.get_dummies(cat_var,drop_first=True,dtype=int)

In [45]:
cat_var.head()

Lead Origin_Landing Page Submission  Lead Origin_Lead Add Form  \
0                                    0                          0   
1                                    0                          0   
2                                    1                          0   
3                                    1                          0   
4                                    1                          0   

   Lead Origin_Lead Import  Last Notable Activity_Email Bounced  \
0                        0                                    0   
1                        0                                    0   
2                        0                                    0   
3                        0                                    0   
4                        0                                    0   

   Last Notable Activity_Email Link Clicked  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   Last Notable Activity_Email Marked Spam  \
0                                        0   
1                                        0   
2                                        0   
3                                        0   
4                                        0   

   Last Notable Activity_Email Opened  Last Notable Activity_Email Received  \
0                                   0                                     0   
1                                   1                                     0   
2                                   1                                     0   
3                                   0                                     0   
4                                   0                                     0   

   Last Notable Activity_Form Submitted on Website  \
0                                                0   
1                                                0   
2                                                0   
3                                                0   
4                                                0   

   Last Notable Activity_Had a Phone Conversation  ...  \
0                                               0  ...   
1                                               0  ...   
2                                               0  ...   
3                                               0  ...   
4                                               0  ...   

   Last Activity_Form Submitted on Website  \
0                                        0   
1                                        0   
2                                        0   
3                                        0   
4                                        0   

   Last Activity_Had a Phone Conversation  \
0                                       0   
1                                       0   
2                                       0   
3                                       0   
4                                       0   

   Last Activity_Olark Chat Conversation  \
0                                      0   
1                                      0   
2                                      0   
3                                      0   
4                                      0   

   Last Activity_Page Visited on Website  \
0                                      1   
1                                      0   
2                                      0   
3                                      0   
4                                      0   

   Last Activity_Resubscribed to emails  Last Activity_SMS Sent  \
0                                     0                       0   
1                                     0                       0   
2                                     0                       0   
3                                     0                       0   
4                                     0                       0

In [46]:
data=pd.concat([num_var,cat_var],axis=1)

In [47]:
data.head()

Do Not Email  Converted  Total Time Spent on Website  \
0             0          0                            0   
1             0          0                          674   
2             0          1                         1532   
3             0          0                          305   
4             0          1                         1428   

   A free copy of Mastering The Interview  TotalVisits  Page Views Per Visit  \
0                                       0          0.0                   0.0   
1                                       0          5.0                   2.5   
2                                       1          2.0                   2.0   
3                                       0          1.0                   1.0   
4                                       0          2.0                   1.0   

   Lead Origin_Landing Page Submission  Lead Origin_Lead Add Form  \
0                                    0                          0   
1                                    0                          0   
2                                    1                          0   
3                                    1                          0   
4                                    1                          0   

   Lead Origin_Lead Import  Last Notable Activity_Email Bounced  ...  \
0                        0                                    0  ...   
1                        0                                    0  ...   
2                        0                                    0  ...   
3                        0                                    0  ...   
4                        0                                    0  ...   

   Last Activity_Form Submitted on Website  \
0                                        0   
1                                        0   
2                                        0   
3                                        0   
4                                        0   

   Last Activity_Had a Phone Conversation  \
0                                       0   
1                                       0   
2                                       0   
3                                       0   
4                                       0   

   Last Activity_Olark Chat Conversation  \
0                                      0   
1                                      0   
2                                      0   
3                                      0   
4                                      0   

   Last Activity_Page Visited on Website  \
0                                      1   
1                                      0   
2                                      0   
3                                      0   
4                                      0   

   Last Activity_Resubscribed to emails  Last Activity_SMS Sent  \
0                                     0                       0   
1                                     0                       0   
2                                     0                       0   
3                                     0                       0   
4                                     0                       0   

   Last Activity_Unreachable  Last Activity_Unsubscribed  \
0                          0                           0   
1                          0                           0   
2                          0                           0   
3                          1                           0   
4                          0                           0   

   Last Activity_View in browser link Clicked  \
0                                           0   
1                                           0   
2                                           0   
3                                           0   
4                                           0   

   Last Activity_Visited Booth in Tradeshow  
0                                         0  
1                                         0  
2                                         0 

In [48]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9074 entries, 0 to 9239
Data columns (total 60 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   Do Not Email                                        9074 non-null   int64  
 1   Converted                                           9074 non-null   int64  
 2   Total Time Spent on Website                         9074 non-null   int64  
 3   A free copy of Mastering The Interview              9074 non-null   int64  
 4   TotalVisits                                         9074 non-null   float64
 5   Page Views Per Visit                                9074 non-null   float64
 6   Lead Origin_Landing Page Submission                 9074 non-null   int64  
 7   Lead Origin_Lead Add Form                           9074 non-null   int64  
 8   Lead Origin_Lead Import                             9074 non-null   int64  
 9   La

In [49]:
data['TotalVisits']=data['TotalVisits'].astype(int)
data['Page Views Per Visit']=data['Page Views Per Visit'].astype(int)

In [50]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9074 entries, 0 to 9239
Data columns (total 60 columns):
 #   Column                                              Non-Null Count  Dtype
---  ------                                              --------------  -----
 0   Do Not Email                                        9074 non-null   int64
 1   Converted                                           9074 non-null   int64
 2   Total Time Spent on Website                         9074 non-null   int64
 3   A free copy of Mastering The Interview              9074 non-null   int64
 4   TotalVisits                                         9074 non-null   int64
 5   Page Views Per Visit                                9074 non-null   int64
 6   Lead Origin_Landing Page Submission                 9074 non-null   int64
 7   Lead Origin_Lead Add Form                           9074 non-null   int64
 8   Lead Origin_Lead Import                             9074 non-null   int64
 9   Last Notable Activity_Em

In [51]:
for co in data.columns:
  print(data[co].value_counts(normalize=True)*100)

Do Not Email
0    92.109323
1     7.890677
Name: proportion, dtype: float64
Converted
0    62.144589
1    37.855411
Name: proportion, dtype: float64
Total Time Spent on Website
0       23.859378
60       0.209389
127      0.198369
75       0.198369
74       0.187348
          ...    
609      0.011020
1035     0.011020
802      0.011020
840      0.011020
927      0.011020
Name: proportion, Length: 1717, dtype: float64
A free copy of Mastering The Interview
0    68.172801
1    31.827199
Name: proportion, dtype: float64
TotalVisits
0      23.815296
2      18.503416
3      14.392771
4      12.342958
5       8.629050
6       5.135552
1       4.353097
7       3.405334
8       2.468592
9       1.807362
10      1.256337
11      0.947763
13      0.528984
12      0.495922
14      0.396738
16      0.231430
15      0.198369
17      0.176328
18      0.165307
20      0.132246
19      0.099184
21      0.066123
23      0.066123
24      0.055102
25      0.055102
27      0.055102
22      0.033061
29   

In [59]:
li=[]
for co in data.columns:
    max_percentage = data[co].value_counts(normalize=True).max() * 100
    if max_percentage <99:
        li.append(co)


In [62]:
li

['Do Not Email',
 'Converted',
 'Total Time Spent on Website',
 'A free copy of Mastering The Interview',
 'TotalVisits',
 'Page Views Per Visit',
 'Lead Origin_Landing Page Submission',
 'Lead Origin_Lead Add Form',
 'Last Notable Activity_Email Link Clicked',
 'Last Notable Activity_Email Opened',
 'Last Notable Activity_Modified',
 'Last Notable Activity_Olark Chat Conversation',
 'Last Notable Activity_Page Visited on Website',
 'Last Notable Activity_SMS Sent',
 'Lead Source_Direct Traffic',
 'Lead Source_Google',
 'Lead Source_Olark Chat',
 'Lead Source_Organic Search',
 'Lead Source_Reference',
 'Lead Source_Referral Sites',
 'Lead Source_Welingak Website',
 'Last Activity_Converted to Lead',
 'Last Activity_Email Bounced',
 'Last Activity_Email Link Clicked',
 'Last Activity_Email Opened',
 'Last Activity_Form Submitted on Website',
 'Last Activity_Olark Chat Conversation',
 'Last Activity_Page Visited on Website',
 'Last Activity_SMS Sent']

In [61]:
# Remove the columns where feature have less importance

In [65]:
data=pd.DataFrame(data,columns=li)

In [66]:
data.shape

(9074, 29)

Scaling of data(fit data into scale)

In [69]:
from sklearn.preprocessing import MinMaxScaler
scaler=MinMaxScaler()
data=pd.DataFrame(scaler.fit_transform(data),columns=data.columns)

In [70]:
data.head()

Do Not Email  Converted  Total Time Spent on Website  \
0           0.0        0.0                     0.000000   
1           0.0        0.0                     0.296655   
2           0.0        1.0                     0.674296   
3           0.0        0.0                     0.134243   
4           0.0        1.0                     0.628521   

   A free copy of Mastering The Interview  TotalVisits  Page Views Per Visit  \
0                                     0.0     0.000000              0.000000   
1                                     0.0     0.019920              0.036364   
2                                     1.0     0.007968              0.036364   
3                                     0.0     0.003984              0.018182   
4                                     0.0     0.007968              0.018182   

   Lead Origin_Landing Page Submission  Lead Origin_Lead Add Form  \
0                                  0.0                        0.0   
1                                  0.0                        0.0   
2                                  1.0                        0.0   
3                                  1.0                        0.0   
4                                  1.0                        0.0   

   Last Notable Activity_Email Link Clicked  \
0                                       0.0   
1                                       0.0   
2                                       0.0   
3                                       0.0   
4                                       0.0   

   Last Notable Activity_Email Opened  ...  Lead Source_Referral Sites  \
0                                 0.0  ...                         0.0   
1                                 1.0  ...                         0.0   
2                                 1.0  ...                         0.0   
3                                 0.0  ...                         0.0   
4                                 0.0  ...                         0.0   

   Lead Source_Welingak Website  Last Activity_Converted to Lead  \
0                           0.0                              0.0   
1                           0.0                              0.0   
2                           0.0                              0.0   
3                           0.0                              0.0   
4                           0.0                              1.0   

   Last Activity_Email Bounced  Last Activity_Email Link Clicked  \
0                          0.0                               0.0   
1                          0.0                               0.0   
2                          0.0                               0.0   
3                          0.0                               0.0   
4                          0.0                               0.0   

   Last Activity_Email Opened  Last Activity_Form Submitted on Website  \
0                         0.0                                      0.0   
1                         1.0                                      0.0   
2                         1.0                                      0.0   
3                         0.0                                      0.0   
4                         0.0                                      0.0   

   Last Activity_Olark Chat Conversation  \
0                                    0.0   
1                                    0.0   
2                                    0.0   
3                                    0.0   
4                                    0.0   

   Last Activity_Page Visited on Website  Last Activity_SMS Sent  
0                                    1.0                     0.0  
1                                    0.0                     0.0  
2                                    0.0                     0.0  
3                                    0.0                     0.0  
4                                    0.0                     0.0  

[5 rows x 29 columns]

In [71]:
from sklearn.model_selection import train_test_split
import statsmodels.api as sm

In [72]:
X=data.drop('Converted',axis=1)
y=data['Converted']

In [73]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=100)

In [74]:
print(X_train.shape ,X_test.shape ,y_train.shape ,y_test.shape)

(6805, 28) (2269, 28) (6805,) (2269,)


In [75]:
X_train=sm.add_constant(X_train)

In [76]:
model = sm.GLM(y_train, X_train, family=sm.families.Binomial())


In [77]:
result=model.fit()

In [78]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:              Converted   No. Observations:                 6805
Model:                            GLM   Df Residuals:                     6776
Model Family:                Binomial   Df Model:                           28
Link Function:                  Logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -2985.5
Date:                Mon, 16 Sep 2024   Deviance:                       5971.0
Time:                        05:13:14   Pearson chi2:                 6.96e+03
No. Iterations:                     7   Pseudo R-squ. (CS):             0.3647
Covariance Type:            nonrobust                                         
=================================================================================================================
                                                    coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------------------
const                                             0.6685      0.444      1.506      0.132      -0.201       1.538
Do Not Email                                     -1.6404      0.195     -8.392      0.000      -2.023      -1.257
Total Time Spent on Website                       4.6280      0.158     29.274      0.000       4.318       4.938
A free copy of Mastering The Interview           -0.0064      0.099     -0.065      0.948      -0.200       0.187
TotalVisits                                      10.4254      2.508      4.156      0.000       5.509      15.342
Page Views Per Visit                             -2.6556      1.200     -2.212      0.027      -5.009      -0.303
Lead Origin_Landing Page Submission              -0.1773      0.098     -1.818      0.069      -0.368       0.014
Lead Origin_Lead Add Form                         1.2537      0.945      1.327      0.185      -0.598       3.106
Last Notable Activity_Email Link Clicked         -1.8219      0.548     -3.325      0.001      -2.896      -0.748
Last Notable Activity_Email Opened               -1.6425      0.390     -4.213      0.000      -2.407      -0.878
Last Notable Activity_Modified                   -2.0172      0.368     -5.482      0.000      -2.738      -1.296
Last Notable Activity_Olark Chat Conversation    -1.9368      0.500     -3.876      0.000      -2.916      -0.957
Last Notable Activity_Page Visited on Website    -1.7893      0.452     -3.956      0.000      -2.676      -0.903
Last Notable Activity_SMS Sent                   -0.5677      0.393     -1.446      0.148      -1.337       0.202
Lead Source_Direct Traffic                       -1.0109      0.366     -2.764      0.006      -1.728      -0.294
Lead Source_Google                               -0.6932      0.358     -1.936      0.053      -1.395       0.009
Lead Source_Olark Chat                            0.2153      0.362      0.595      0.552      -0.493       0.924
Lead Source_Organic Search                       -0.9138      0.369     -2.473      0.013      -1.638      -0.190
Lead Source_Reference                             2.0121      0.933      2.156      0.031       0.183       3.842
Lead Source_Referral Sites                       -0.9960      0.458     -2.173      0.030      -1.895      -0.097
Lead Source_Welingak Website                      3.6081      1.163      3.103      0.002       1.329       5.887
Last Activity_Converted to Lead                  -0.9604      0.353     -2.718      0.007      -1.653      -0.268
Last Activity_Email Bounced                      -1.4527      0.450     -3.225      0.001      -2.336      -0.570
Last Activity_Email Link Clicked                 -0.3919      0.446     -0.879      0.380      -1.266       0.482
Last 

In [83]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

In [84]:
print(confusion_matrix(y_train,result.predict(X_train).round()))

[[3679  519]
 [ 841 1766]]


In [85]:
print(accuracy_score(y_train,result.predict(X_train).round()))

0.8001469507714916


In [87]:
print(classification_report(y_train,result.predict(X_train).round()))

              precision    recall  f1-score   support

         0.0       0.81      0.88      0.84      4198
         1.0       0.77      0.68      0.72      2607

    accuracy                           0.80      6805
   macro avg       0.79      0.78      0.78      6805
weighted avg       0.80      0.80      0.80      6805



In [93]:
# y_pred
X_test=sm.add_constant(X_test)
y_pred = result.predict(X_test)

In [94]:
y_test = y_test.astype(int)

In [95]:
y_test_pred_final = pd.DataFrame({'actual':y_test.values, 'predicted':y_pred})

y_test_pred_final.head()

actual  predicted
3184       0   0.182662
1435       1   0.680385
7788       0   0.160571
4121       1   0.857303
3737       0   0.128367

In [96]:
y_test_pred_final['predicted'] = y_test_pred_final.predicted.map(lambda x: 1 if x > 0.3 else 0)

# Let's see the head
y_test_pred_final.head()

actual  predicted
3184       0          0
1435       1          1
7788       0          0
4121       1          1
3737       0          0

In [97]:
print(confusion_matrix(y_test,y_test_pred_final['predicted']))
print(accuracy_score(y_test,y_test_pred_final['predicted']))
print(classification_report(y_test,y_test_pred_final['predicted']))

[[1118  323]
 [ 151  677]]
0.7910973997355664
              precision    recall  f1-score   support

           0       0.88      0.78      0.83      1441
           1       0.68      0.82      0.74       828

    accuracy                           0.79      2269
   macro avg       0.78      0.80      0.78      2269
weighted avg       0.81      0.79      0.79      2269



In [98]:
coefficients = result.params
print(coefficients.sort_values(ascending=False))

TotalVisits                                      10.425411
Total Time Spent on Website                       4.627988
Lead Source_Welingak Website                      3.608053
Lead Source_Reference                             2.012142
Lead Origin_Lead Add Form                         1.253685
const                                             0.668549
Last Activity_SMS Sent                            0.229124
Lead Source_Olark Chat                            0.215258
A free copy of Mastering The Interview           -0.006421
Last Activity_Email Opened                       -0.124106
Lead Origin_Landing Page Submission              -0.177286
Last Activity_Email Link Clicked                 -0.391873
Last Activity_Form Submitted on Website          -0.445507
Last Activity_Page Visited on Website            -0.493275
Last Notable Activity_SMS Sent                   -0.567668
Lead Source_Google                               -0.693155
Lead Source_Organic Search                       -0.9137

In [99]:
# prompt: 1.	Which are the top three variables in your model which contribute most towards the probability of a lead getting converted? according this model

# Get the coefficients from the model
coefficients = result.params

# Sort the coefficients in descending order
sorted_coefficients = coefficients.sort_values(ascending=False)

# Get the top three variables
top_three_variables = sorted_coefficients.index[:3]

# Print the top three variables
print(top_three_variables)


Index(['TotalVisits', 'Total Time Spent on Website',
       'Lead Source_Welingak Website'],
      dtype='object')


# Our model is gives 79 % accurate score on Test dataset